README.md
# tcptunnel
用于两种场景：
1. 直接的端口转发，这个好理解
2. 做内网服务器到公网的映射访问，用于解决内网服务器没有公网IP或者无法进行端口映射的场景

想要完成ngrok和lcx等类似的功能，对于lcx定义的slave啊，listen啊，tran啊我觉得很大歧义，半天理解不了。所以我发明了三个简单易懂的连接方式：
* 公网服务器： publicserver，用于做转发的，监听一个对外开放的端口就行（对，我这里只要一个端口就行）
* 内网服务器：natserver，也就是实际希望被外网访问的局域网服务器（或者是局域网代理）
* 客户端：client，连接客户端，这个很好理解，实际的访问者，本地启动后，通过其他客户端连接本地监听的端口就相当于访问内网服务器

是不是更容易理解？是的话直接夸我。

**注意：没有经过大量实际测试，请谨慎用于生产环境。**

## 编译
由于用到了1.9才有的sync.Map, 所以编译环境必须是1.9+，见谅见谅 ；）
```
git clone https://github.com/LubyRuffy/tcptunnel
go get
go build
```
生成tcptunnel文件

跨平台编译，比如到NAS服务器或者树莓派等ARM平台，执行：
```
GOOS=linux GOARCH=arm GOARM=5 go build
```

## 运行
直接执行是读取config.toml配置文件中的内容，最主要的是Mode和对应的配置内容，后续在配置文件中说明。
```
./tcptunnel 
```

### 作为内网映射运行：
- 作为publciserver执行，放到公网服务器
```
./tcptunnel -m publicserver
```
- 作为natserver执行，放到内网的服务器
```
./tcptunnel -m natserver
```
- 作为client执行，放到需要访问内网服务器的客户端
```
./tcptunnel -m client

然后连接本地端口就相当于连接natserver里面对应的服务器了

```
natserver 和 client 通信是通过约定好一致的唯一ID来进行。

### 作为tcpproxy执行，也就是端口转发
```
./tcptunnel -m tcpproxy
```

## 配置文件说明
默认读取config.toml文件，
```
# 模式： 支持publicserver，natserver，client，tcpproxy。可以通过命令行的-m参数覆盖
Mode = "publicserver"

# 连接公网服务器的地址，格式为 host:port
# 在Mode为 natserver 和 clientconnect 时有效
PublicServerAddr = "127.0.0.1:10011"

# 端口转发模式，仅仅在Mode为 tcpproxy 时有效
[TcpProxies]
    # 数组，可以多个映射关系
    [TcpProxies.proxy80]
    LocalBindAddr = "127.0.0.1:1234"
    RemoteServerAddr = "192.168.1.1:80"
    Type = "http"

    [TcpProxies.proxy22]
    LocalBindAddr = "127.0.0.1:1235"
    RemoteServerAddr = "192.168.1.1:22"

# 公网服务器监听的地址，仅仅在Mode为 publicserver 时有效，格式为 ip:port
[PublicServer]
LocalBindAddr = "127.0.0.1:10011"

# 端口转发模式，仅仅在Mode为 natserver 时有效
[NatServer]
    # 数组，可以多个映射关系，ID用于注册，客户端连接的时候直接通过ID来进行查找
    [NatServer.test]
    RemoteServerAddr = "192.168.1.1:80"
    ID = "test"
    Type = "http"

    [NatServer.test1]
    RemoteServerAddr = "192.168.1.1:22"
    ID = "test1"
    
# 端口转发模式，仅仅在Mode为 client 时有效
[ClientConnect]
    # 数组，可以多个映射关系，ID用于标示连接时指定NAT后的服务器对象
    [ClientConnect.test]
    LocalBindAddr = "127.0.0.1:1234"
    ID = "test"
```

## 流程说明
```
natserver    REGISTER   -> publicserver
                           publicserver <- CONNECT client <- application tcp connect
natserver NEWDATASTREAM <- publicserver
natserver   DATASTREAM  -> publicserver
                           publicserver -> 200 OK client <-> application tcp connect
```

HISTORY
v0.1 - Jan 13, 2018
alpha

v1.0 - Feb 9, 2018
稳定版本


config.go
package main

import (
	"flag"
	"fmt"
	// _ "net/http/pprof"
	"os"
	"runtime/pprof"

	"github.com/BurntSushi/toml"
	//"go/types"
)

type TcpProxyConfig struct {
	LocalBindAddr    string
	RemoteServerAddr string
	Type             string
}

type PublicServerConfig struct {
	LocalBindAddr string
}

type ClientConnectConfig struct {
	LocalBindAddr string
	ID            string
}

type NatServerConfig struct {
	RemoteServerAddr string
	ID               string
	Type             string
}

type TomlConfig struct {
	Mode             string
	PublicServerAddr string
	TcpProxies       map[string]TcpProxyConfig
	PublicServer     PublicServerConfig
	NatServer        map[string]NatServerConfig
	ClientConnect    map[string]ClientConnectConfig
}

var configOptions TomlConfig

func init() {
	cpuProfile := flag.String("p", "", "write cpu profile to file")
	configFile := flag.String("c", "config.toml", "config file")
	mode := flag.String("m", "", "mode, can overwrite config_file's setting")
	flag.Parse()

	if *configFile == "" {
		panic("config file is not specified")
	}

	if _, err := toml.DecodeFile(*configFile, &configOptions); err != nil {
		// handle error
		panic(err)
	}

	if *mode != "" {
		configOptions.Mode = *mode
	}

	if *cpuProfile != "" {
		// go func() {
		// 	log.Println(http.ListenAndServe("localhost:6060", nil))
		// }()

		f, err := os.Create(*cpuProfile)
		if err != nil {
			fmt.Println(err)
		}
		pprof.StartCPUProfile(f)
		defer pprof.StopCPUProfile()
	}
}


iobind.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"log"
	"math/rand"
	"net"
	"net/http"
	"sync"

	"github.com/xtaci/smux"
)

type onConnection func(net.Conn)
type onStream func(*smux.Stream)

func ioCopy(dst io.ReadWriter, src io.ReadWriter) (err error) {
	buf := make([]byte, 32*1024)
	n := 0
	for {
		n, err = src.Read(buf)
		if n > 0 {
			if _, e := dst.Write(buf[0:n]); e != nil {
				return e
			}
		}
		if err != nil {
			return
		}
	}
}

func IoBind(dst io.ReadWriteCloser, src io.ReadWriteCloser, fn func(err interface{})) {
	defer func() {
		if err := recover(); err != nil {
			log.Printf("bind crashed %s", err)
		}
	}()
	e1 := make(chan interface{}, 1)
	e2 := make(chan interface{}, 1)
	go func() {
		defer func() {
			if err := recover(); err != nil {
				log.Printf("bind crashed %s", err)
			}
		}()
		//_, err := io.Copy(dst, src)
		err := ioCopy(dst, src)
		e1 <- err
	}()
	go func() {
		defer func() {
			if err := recover(); err != nil {
				log.Printf("bind crashed %s", err)
			}
		}()
		//_, err := io.Copy(src, dst)
		err := ioCopy(src, dst)
		e2 <- err
	}()
	var err interface{}
	select {
	case err = <-e1:
		//log.Printf("e1")
	case err = <-e2:
		//log.Printf("e2")
	}
	src.Close()
	dst.Close()

	fn(err)
}

func random(min int, max int) string {
	return fmt.Sprintf("%v", rand.Intn(max-min)+min)
}

// 通用TCP监听流程
func listenTCPServer(wg *sync.WaitGroup, addr string, fn onConnection) (err error) {
	defer func() {
		wg.Done()
	}()

	var listener net.Listener
	listener, err = net.Listen("tcp", addr)
	if err != nil {
		return err
	}
	log.Printf("bind server to %s ok, listenning...\n", addr)
	for {
		var conn net.Conn
		conn, err = listener.Accept()
		if err != nil {
			log.Printf("Accept error, %v", err)
		}

		go func(newconn net.Conn) {
			fn(newconn)
		}(conn)
	}
}

func handleRequest(conn net.Conn, outConn net.Conn, inReq *http.Request, remoteAddr string) (resp *http.Response, err error) {
	if inReq == nil {
		return
	}

	outReq := new(http.Request)
	*outReq = *inReq // includes shallow copies of maps, but we handle this in Director
	outReq.Host = remoteAddr
	err = outReq.Write(outConn)
	if err != nil {
		log.Printf("[http] %s - %s : %s", conn.RemoteAddr(), conn.LocalAddr(), err)
		return
	}

	resp, err = http.ReadResponse(bufio.NewReader(outConn), outReq)
	if err != nil {
		log.Printf("[http] %s - %s : %s", conn.RemoteAddr(), conn.LocalAddr(), err)
		return
	}
	return
}

func HTTPBind(conn net.Conn, outConn net.Conn, remoteAddr string, id string) {
	defer func() {
		conn.Close()
		outConn.Close()
	}()
	for {
		req, err := http.ReadRequest(bufio.NewReader(conn))
		if err != nil {
			log.Printf("%s [http ReadRequest from conn] %s - %s : %s", id, conn.RemoteAddr(), conn.LocalAddr(), err)
			return
		}

		// log.Printf("%s [http] %s - %s : %v -> ", id, conn.RemoteAddr(), conn.LocalAddr(), req.URL)

		resp, err := handleRequest(conn, outConn, req, remoteAddr)
		if err != nil {
			log.Printf("%s [http handleRequest] %s - %s : %s", id, conn.RemoteAddr(), conn.LocalAddr(), err)
			return
		}

		// log.Printf("%s [http] %s - %s : %v -> %s <-", id, conn.RemoteAddr(), conn.LocalAddr(), req.URL, resp.Status)

		err = resp.Write(conn)
		if err != nil {
			log.Printf("%s [http Write to conn] %s - %s : %s", id, conn.RemoteAddr(), conn.LocalAddr(), err)
			return
		}

		// log.Printf("%s [http finished] %s - %s : %v -> %s <-", id, conn.RemoteAddr(), conn.LocalAddr(), req.URL, resp.Status)

		return
	}
}


tcpproxy.go
package main

import (
	"fmt"
	"io"
	"log"
	"net"
	"sync"
)

func tcpProxyPair(newconn net.Conn, v TcpProxyConfig) {
	defer func() {
		newconn.Close()
	}()

	outConn, err := net.Dial("tcp", v.RemoteServerAddr)
	if err != nil {
		panic(err)
		return
	}

	defer func() {
		outConn.Close()
	}()

	if v.Type == "http" {
		HTTPBind(newconn, outConn, v.RemoteServerAddr, "-")
	} else {
		IoBind(newconn, outConn, func(err interface{}) {
			if err != io.EOF && err != nil {
				log.Printf("IoBind failed: %v\n", err)
			}

			inAddr := newconn.RemoteAddr().String()
			outAddr := outConn.RemoteAddr().String()
			log.Printf("newconn %s - %s released", inAddr, outAddr)
		})
	}
}

// 端口转发
func createOneTcpProxy(wg *sync.WaitGroup, v TcpProxyConfig) {
	defer func() {
		if err := recover(); err != nil {
			log.Printf("bind crashed : %s", err)
		}
		wg.Done()
	}()

	l, err := net.Listen("tcp", v.LocalBindAddr)
	if err == nil {
		log.Printf("bind server to %s ok, listenning...\n", v.LocalBindAddr)
		for {
			var conn net.Conn
			conn, err = l.Accept()
			if err == nil {
				go tcpProxyPair(conn, v)
			} else {
				panic(err)
				return
			}
		}
	} else {
		panic("bind error")
	}
}

func tcpProxy() {
	wg := sync.WaitGroup{}
	for _, v := range configOptions.TcpProxies {
		fmt.Printf("proxy of %s -> %s \n", v.LocalBindAddr, v.RemoteServerAddr)
		go createOneTcpProxy(&wg, v)
		wg.Add(1)
	}
	wg.Wait()
}


natserver.go
package main

import (
	"bufio"
	"errors"
	"fmt"
	"io"
	"log"
	"net"
	"sync"
	"time"

	"net/http"

	"github.com/xtaci/smux"
)

func doNewDataStream(id string, stream *smux.Stream, v NatServerConfig) {
	defer func() {
		log.Printf("close stream id: %v\n", id)
		stream.Close()
	}()

	buf := make([]byte, 65536)

	_, err := stream.Write([]byte(fmt.Sprintf("DATASTREAM %s HTTP/1.0\r\n\r\n", id)))
	if err != nil {
		log.Println(fmt.Sprintf("stream.Write failed: %v", err))
		return
	}

	_, err = stream.Read(buf)
	if err != nil {
		log.Printf("stream.Read failed: %v\n", err)
		return
	}
	//log.Printf("%s stream.Read : %s\n", id, string(buf[:n]))
	// log.Printf("%s try to net.Dial to %s\n", id, v.RemoteServerAddr)

	outConn, err := net.Dial("tcp", v.RemoteServerAddr)
	if err != nil {
		log.Printf("%s net.Dial failed: %v", id, err)
		return
	}
	// log.Printf("%s net.Dial to %s ok\n", id, v.RemoteServerAddr)

	if v.Type == "http" {
		HTTPBind(stream, outConn, v.RemoteServerAddr, id)
	} else {
		IoBind(stream, outConn, func(err interface{}) {
			if err != io.EOF && err != nil {
				log.Printf("IoBind failed: %v\n", err)
			}

			// inAddr := stream.RemoteAddr().String()
			// outAddr := outConn.RemoteAddr().String()
			// log.Printf("conn %s - %s released", inAddr, outAddr)
		})
	}

}

func connectOneServer(publicServerAddr string, v NatServerConfig) (err error) {
	cli, err := net.DialTimeout("tcp", publicServerAddr, time.Second*10)
	if err != nil {
		return errors.New(fmt.Sprintf("Dial to server %s failed: %v!", publicServerAddr, err))
	}

	log.Println("connected to public server : ", publicServerAddr)

	defer func() {
		cli.Close()
	}()

	session, _ := smux.Client(cli, nil)
	stream, _ := session.OpenStream()

	defer func() {
		stream.Close()
		session.Close()
	}()

	buf := make([]byte, 65536)

	stream.Write([]byte(fmt.Sprintf("REGISTER /%s HTTP/1.0\r\n\r\n", v.ID)))
	n, err := stream.Read(buf)
	if err != nil {
		return errors.New(fmt.Sprintf("stream.Read failed: %v", err))
	}
	log.Println(string(buf[:n]))

	var req *http.Request
	for {
		req, err = http.ReadRequest(bufio.NewReader(stream))
		if err != nil {
			return errors.New(fmt.Sprintf("recvReq failed: %v", err))
		}

		switch req.Method {
		case "NEWDATASTREAM":
			// stream.Write([]byte("200 OK\r\n\r\n"))
			log.Println("NEWDATASTREAM", req.RequestURI)
			newstream, _ := session.OpenStream()
			go doNewDataStream(req.RequestURI, newstream, v)
		default:
			log.Println("Unknown msg type")
		}
	}

	return
}

// 控制stream通道
func connectServer(wg *sync.WaitGroup, publicServerAddr string, v NatServerConfig) {
	defer func() {
		wg.Done()
	}()

	var err error
	for {
		err = connectOneServer(publicServerAddr, v)
		if err != nil {
			log.Println(err)
			time.Sleep(time.Second * 5)
			continue
		}
		break
	}
}

func natServer() {
	wg := sync.WaitGroup{}
	for _, v := range configOptions.NatServer {
		fmt.Printf("ID %s -> %s \n", v.ID, v.RemoteServerAddr)
		go connectServer(&wg, configOptions.PublicServerAddr, v)
		wg.Add(1)
	}
	wg.Wait()
}


clientconnect.go
package main

import (
	"fmt"
	"io"
	"log"
	"net"
	"strings"
	"sync"
	"time"

	"github.com/xtaci/smux"
)

type ControlSession struct {
	Session *smux.Session
}

// 控制stream通道
func getControlSession(addr string) (session *smux.Session, err error) {
	cli, err := net.Dial("tcp", addr)
	if err != nil {
		return
	}
	session, err = smux.Client(cli, nil)
	if err != nil {
		return
	}
	return
}

func bindConnToServer(id string, newconn net.Conn, session *ControlSession) {
	var stream *smux.Stream
	var err error

	defer func() {
		newconn.Close()
		if stream != nil {
			stream.Close()
		}
	}()

	if session.Session == nil {
		log.Println("[ERROR] session.Session not connected!")
		return
	}

	buf := make([]byte, 1024*1024)
	stream, err = session.Session.OpenStream()

	if err != nil {
		log.Println("session.OpenStream error:", err)
		return
	}
	log.Println("session.OpenStream ok, id is :", stream.ID())

	n, err := stream.Write([]byte(fmt.Sprintf("CONNECT /%s HTTP/1.0\r\n\r\n", id)))
	if err != nil {
		log.Println("stream.Write CONNECT error:", err)
		return
	}

	// 10秒读取超时
	stream.SetReadDeadline(time.Now().Add(time.Duration(10) * time.Second))
	n, err = stream.Read(buf)
	if err != nil {
		log.Println("stream.Read error:", err)
		return
	}

	log.Println("stream.Read size : ", n)
	if strings.Contains(string(buf[:n]), "200 OK") {
		log.Printf("CONNECT to server OK\n")

		stream.SetReadDeadline(time.Time{})
		IoBind(newconn, stream, func(err interface{}) {
			if err != io.EOF && err != nil {
				log.Printf("IoBind failed: %v\n", err)
			}
		})
	} else {
		log.Printf("CONNECT to server failed: %s\n", string(buf[:n]))
		return
	}
}

func clientConnect() {
	session := ControlSession{}
	var err error

	go func(ctlsession *ControlSession) {
		// 获取控制session
		for {
			if ctlsession.Session != nil && !ctlsession.Session.IsClosed() {
				time.Sleep(time.Second * 5)
			} else {
				ctlsession.Session, err = getControlSession(configOptions.PublicServerAddr)
				if err != nil {
					log.Printf("connect to server failed: %v, retry after 5 seconds\n", err)
					time.Sleep(time.Second * 5)
					continue
				}
			}
		}
	}(&session)

	wg := sync.WaitGroup{}

	// 监听
	for _, v := range configOptions.ClientConnect {

		go func(config ClientConnectConfig) {
			log.Println("bind ", config.LocalBindAddr, " to ", config.ID)
			listenTCPServer(&wg, config.LocalBindAddr, func(newconn net.Conn) {
				bindConnToServer(config.ID, newconn, &session)
			})
		}(v)
		wg.Add(1)
	}

	wg.Wait()
}


publicserver.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"log"
	"math/rand"
	"net"
	"net/http"
	"sync"
	"time"

	"github.com/xtaci/smux"
)

type ControlStream struct {
	sync.Mutex
	Stream *smux.Stream
}

var natCtlMap sync.Map     // 控制流
var dataStreamMap sync.Map // 数据通道流

// 公网监听服务器
func listenPublicServer(wg *sync.WaitGroup, addr string, fn onStream) (err error) {
	err = listenTCPServer(wg, addr, func(newconn net.Conn) {

		sess, err := smux.Server(newconn, nil)
		if err != nil {
			log.Printf("smux should not fail with default config: %v", err)
			return
		}

		// 断开连接后取消所有ID的注册 https://github.com/LubyRuffy/tcptunnel/issues/3
		session_streams := make([]*smux.Stream, 0)
		defer func() {
			for _, s := range session_streams {
				natCtlMap.Range(func(k, v interface{}) bool {
					if s == v.(*ControlStream).Stream {
						log.Printf("unregister ID: %v", k)
						natCtlMap.Delete(k)
					}
					return true
				})
			}
		}()

		for {
			stream, err := sess.AcceptStream()
			if err != nil {
				log.Printf("AcceptStream failed: %v", err)
				return
			}
			// log.Printf("AcceptStream ok, id: %v", stream.ID)

			go func(newstream *smux.Stream) {
				fn(newstream)
			}(stream)
			session_streams = append(session_streams, stream)
		}
	})
	return
}

func doConnect(req *http.Request, stream *smux.Stream) {
	var err error
	// buf := make([]byte, 1024*1024)

	if ctlStream, ok := natCtlMap.Load(req.RequestURI); ok {
		dataStreamID := "/" + random(1000, 10000000)
		// log.Println(" ctlStream: ", ctlStream, " dataStreamID: ", dataStreamID)

		// 新建等待事件
		dataStreamMap.Store(dataStreamID, make(chan *smux.Stream, 1))
		// log.Println("wait DATASTREAM event:", dataStreamID)
		defer func() {
			dataStreamMap.Delete(dataStreamID)
		}()

		// 通知建立新的数据通道
		log.Println("send NEWDATASTREAM", dataStreamID)
		ctlStream.(*ControlStream).Lock()
		_, err = ctlStream.(*ControlStream).Stream.Write([]byte(fmt.Sprintf("NEWDATASTREAM %s HTTP/1.0\r\n\r\n", dataStreamID)))
		time.Sleep(time.Millisecond * 20) // 这里必须延时一下，否则在浏览器并发连接的时候，natserver在readRequest的过程中会接受不到数据，这个需要再想办法，目前临时加一个延时
		ctlStream.(*ControlStream).Unlock()
		if err != nil {
			log.Println(" stream write data error: ", err)
		}

		// 等待通知
		waitChan, ok := dataStreamMap.Load(dataStreamID)
		if !ok {
			panic(fmt.Sprintf(" dataStreamMap not found of : %s", dataStreamID))
		}

		// dataStream := <-waitChan.(chan *smux.Stream)
		// log.Println("DATASTREAM event ok: ", dataStream)
		var dataStream *smux.Stream
		select {
		case dataStream = <-waitChan.(chan *smux.Stream):
			// log.Println("DATASTREAM event ok: ", dataStream)
			dataStreamMap.Delete(dataStreamID)
		case <-time.After(time.Second * 10):
			log.Println("waitChan DATASTREAM timeout: ", dataStreamID)
			dataStreamMap.Delete(dataStreamID)
			return
		}

		// log.Println("DUMP 222222")
		// dataStreamMap.Range(func(key, value interface{}) bool {
		// 	log.Println(key, value)
		// 	return true
		// })
		// log.Println("222222 DUMP")

		stream.Write([]byte("200 OK\r\n\r\n"))

		IoBind(dataStream, stream, func(err interface{}) {
			if err != io.EOF && err != nil {
				log.Printf("IoBind failed: %v\n", err)
			}
		})
	} else {
		log.Println("Could not found stream of id : ", req.RequestURI)

		// https://github.com/LubyRuffy/tcptunnel/issues/2
		stream.Write([]byte("404 ID Not Found\r\n\r\n"))
		stream.Close()
	}
}

func doDataStream(req *http.Request, stream *smux.Stream) {
	log.Println("DATASTREAM msg to ", req.RequestURI)
	stream.Write([]byte("200 OK\r\n\r\n"))
	waitChan, ok := dataStreamMap.Load(req.RequestURI)
	if ok {
		waitChan.(chan *smux.Stream) <- stream
	} else {
		log.Println(" dataStreamMap not found of : ", req.RequestURI)
	}
}

func publicServer() {
	rand.Seed(time.Now().UnixNano())
	wg := sync.WaitGroup{}

	listenPublicServer(&wg, configOptions.PublicServer.LocalBindAddr, func(stream *smux.Stream) {

		req, err := http.ReadRequest(bufio.NewReader(stream))
		if err != nil {
			log.Printf("recv request failed: %v", err)
			return
		}

		// log.Println("recv req: ", req)
		switch req.Method {
		case "CONNECT":
			// log.Println("CONNECT msg to ", req.RequestURI)
			go doConnect(req, stream)
		case "REGISTER":
			log.Println("REGISTER msg to ", req.RequestURI)

			// 排他唯一性，https://github.com/LubyRuffy/tcptunnel/issues/3
			if _, ok := natCtlMap.Load(req.RequestURI); ok {
				log.Println("ID Already Registered", req.RequestURI, ", reject it!")
				stream.Write([]byte("500 ID Already Registered\r\n\r\n"))
				stream.Close()
				return
			} else {
				ctlStream := ControlStream{Stream: stream}
				natCtlMap.Store(req.RequestURI, &ctlStream)
				stream.Write([]byte("200 OK\r\n\r\n"))
			}

			// log.Println("DUMP====")
			// natCtlMap.Range(func(key, value interface{}) bool {
			// 	log.Println(key, value)
			// 	return true
			// })
			// log.Println("====DUMP")
		case "DATASTREAM":
			go doDataStream(req, stream)

		default:
			log.Println("Unknown msg type")
			return
		}
	})
}
